### PR TITLE
adding functions to manage life cycle of resources in ITComputeTest

### PR DESCRIPTION
**Pull request by neozwu**
In ITComputeTest, we use "JUnit timeout rule" to specify test case timeout. Timeout rule (unlike timeout parameter in @Test annotation) applies to the entire test fixture (that includes all @Before methods + @Test method + all @After methods). Therefore, when a test times out in the middle of execution, teardown method (@After method) won't be executed. This will result in resource leak, and accumulation of unused resources (instances, disks, snapshots, etc.) in google cloud compute engine.  One approach is to use timeout parameter rather than timeout rule and properly clean up resources in @After method.  However timeout parameter will force tests (@Test method) to run in a different thread, thus may cause potential thread-safety issue.  In this PR, a different approach is used, which registers all resources that are created by test cases in a static data structure, and try to delete all these resources after all tests are executed (i.e. in the @Afterclass method).  The implementation tries to be as generic as possible, but without introducing any change of existing resource objects.

### CODE REVIEW

File: google-cloud-compute/src/main/java/com/google/cloud/compute/testing/ManagedResourceHelper.java [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+import com.google.cloud.compute.Operation;
+import com.google.cloud.compute.Snapshot;
+import com.google.cloud.compute.SnapshotId;
+import com.google.cloud.compute.Subnetwork;
+import com.google.cloud.compute.SubnetworkId;
+import com.google.common.base.Function;
+import java.util.ArrayList;
+import java.util.List;
+
+public class ManagedResourceHelper {
```
Comment (shinfan) [inline google-cloud-compute/src/main/java/com/google/cloud/compute/testing/ManagedResourceHelper.java]: Why is this only for Compute? It seems to me that this helper can be generalized and used for other APIs as well. e.g. We can make this helper a general class and create a API-specific resource deletion handler per API.

File: google-cloud-compute/src/main/java/com/google/cloud/compute/testing/ManagedResourceHelper.java [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+    final T res;
+    Function<T, Void> func;
+
+    private ManagedResource(T res) {
+      this.res = res;
+
+      @SuppressWarnings("unchecked")
+      Function<T, Void> localFunc = (Function<T, Void>) getDeleteMethod(res);
+
+      this.func = localFunc;
```
Comment (shinfan) [inline google-cloud-compute/src/main/java/com/google/cloud/compute/testing/ManagedResourceHelper.java]: this.func = (Function<T, Void>) getDeleteMethod(res);

File: google-cloud-compute/src/main/java/com/google/cloud/compute/testing/ManagedResourceHelper.java [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+  public <T> ManagedResourceHelper addManagedResource(T res) {
+    ManagedResource<T> r = new ManagedResource<>(res);
+    resources.add(r);
+    return this;
+  }
+
+  public void cleanUpResources() {
+    for (ManagedResource<?> r : resources) {
+      r.delete();
+      //TODO (neowu): add logging if a resource has not been deleted in test fixture
```
Comment (shinfan) [inline google-cloud-compute/src/main/java/com/google/cloud/compute/testing/ManagedResourceHelper.java]: How do you detect if a resource has been deleted or not?

**Comment by shinfan** [comment]
Thanks for fixing this! Just a few comments.

File: google-cloud-compute/src/main/java/com/google/cloud/compute/testing/ManagedResourceHelper.java [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+      return new Function<Address, Void>() {
+        @Override
+        public Void apply(Address address) {
+          try {
+            Operation operation = address.delete();
+            operation.waitFor();
+          } catch (Exception e) {
+
+          } finally {
+            return null;
```
Comment (garrettjonesgoogle) [inline google-cloud-compute/src/main/java/com/google/cloud/compute/testing/ManagedResourceHelper.java]: It doesn't make sense for this `return` to be in a `finally` block.

File: google-cloud-compute/src/main/java/com/google/cloud/compute/testing/ManagedResourceHelper.java [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+
+  private <T> Function<?, Void> getDeleteMethod(T res) {
+    if (res instanceof Address) {
+      return new Function<Address, Void>() {
+        @Override
+        public Void apply(Address address) {
+          try {
+            Operation operation = address.delete();
+            operation.waitFor();
+          } catch (Exception e) {
```
Comment (garrettjonesgoogle) [inline google-cloud-compute/src/main/java/com/google/cloud/compute/testing/ManagedResourceHelper.java]: We should print the exceptions that we see in deletes.

File: google-cloud-compute/src/main/java/com/google/cloud/compute/testing/ManagedResourceHelper.java [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+  public void cleanUpResources() {
+    for (ManagedResource<?> r : resources) {
+      r.delete();
+      //TODO (neowu): add logging if a resource has not been deleted in test fixture
+    }
+    resources.clear();
+  }
+
+  private <T> Function<?, Void> getDeleteMethod(T res) {
+    if (res instanceof Address) {
```
Comment (garrettjonesgoogle) [inline google-cloud-compute/src/main/java/com/google/cloud/compute/testing/ManagedResourceHelper.java]: With this setup, we won't know that a particular type isn't supported at compile time, yet we know the type at the call site of `addManagedResource`. I think instead we should have a separate `addManagedResource` overload for each type in Compute. Then, if we add a new one, we'll know at compile time if we don't support it yet.

File: google-cloud-compute/src/main/java/com/google/cloud/compute/testing/ManagedResourceHelper.java [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+
+      this.func = localFunc;
+    }
+
+    public void delete() {
+      func.apply(res);
+    }
+  }
+
+  public <T> ManagedResourceHelper addManagedResource(T res) {
```
Comment (garrettjonesgoogle) [inline google-cloud-compute/src/main/java/com/google/cloud/compute/testing/ManagedResourceHelper.java]: you could probably just shorten this to `add`.

File: google-cloud-compute/src/main/java/com/google/cloud/compute/testing/ManagedResourceHelper.java (hunk 1 above)
Comment (garrettjonesgoogle) [inline google-cloud-compute/src/main/java/com/google/cloud/compute/testing/ManagedResourceHelper.java]: I think the `helper` suffix is too general. Maybe name this class `ResourceCleaner`.

File: google-cloud-compute/src/main/java/com/google/cloud/compute/testing/ManagedResourceHelper.java [hunk 8]
```diff
@@ -0,0 +1,230 @@
+/*
+ * Copyright 2016 Google Inc. All Rights Reserved.
```
Comment (neozwu) [inline google-cloud-compute/src/main/java/com/google/cloud/compute/testing/ManagedResourceHelper.java]: resolved

File: google-cloud-compute/src/main/java/com/google/cloud/compute/testing/ManagedResourceCleaner.java [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+  private <T> Function<?, Void> getDeleteMethod(T res) {
+
+    if (res instanceof AddressId) {
+      return new Function<AddressId, Void>() {
+        @Override
+        public Void apply(AddressId addressId) {
+          compute.deleteAddress(addressId);
+          return null;
+        }
+      };
```
Comment (garrettjonesgoogle) [inline google-cloud-compute/src/main/java/com/google/cloud/compute/testing/ManagedResourceCleaner.java]: It would be much cleaner to put these delete `Function`s inside the corresponding `add` method for each type. Then, you never have to bother with an `instanceof` check. Additionally, I think it would mean you could remove the type parameterization (`<T>`) from MonitoredResource too.

File: google-cloud-compute/src/main/java/com/google/cloud/compute/testing/ManagedResourceHelper.java (hunk 1 above)
Comment (neozwu) [inline google-cloud-compute/src/main/java/com/google/cloud/compute/testing/ManagedResourceHelper.java]: I agree the problem here is more generic and not only applied to compute engine.  I'd like to have a specific version for compute engine for now, and create a more generic version if we need. The reason is if we make a more general class, I might need to change things in various libraries.

File: google-cloud-compute/src/main/java/com/google/cloud/compute/testing/ManagedResourceHelper.java (hunk 2 above)
Comment (neozwu) [inline google-cloud-compute/src/main/java/com/google/cloud/compute/testing/ManagedResourceHelper.java]: yeah, this is ugly. The reason I'd like to have a seemingly repeated statement is because then I can limit the scope of `@SuppressWarnings("unchecked")`

File: google-cloud-compute/src/main/java/com/google/cloud/compute/testing/ManagedResourceCleaner.java [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+  public ManagedResourceCleaner add(AddressId addressId) {
+    ManagedResource<AddressId> r = new ManagedResource<>(addressId);
+    resources.put(addressId, r);
+    return this;
+  }
+
+  public ManagedResourceCleaner remove(AddressId addressId) {
+    resources.remove(addressId);
+    return this;
+  }
```
Comment (garrettjonesgoogle) [inline google-cloud-compute/src/main/java/com/google/cloud/compute/testing/ManagedResourceCleaner.java]: You don't need an overload for `remove` - you can have just one implementation. Maybe throw an exception if the id is not found in `resources`.

File: google-cloud-compute/src/main/java/com/google/cloud/compute/testing/ManagedResourceHelper.java (hunk 1 above)
Comment (neozwu) [inline google-cloud-compute/src/main/java/com/google/cloud/compute/testing/ManagedResourceHelper.java]: done

File: google-cloud-compute/src/main/java/com/google/cloud/compute/testing/ManagedResourceHelper.java (hunk 7 above)
Comment (neozwu) [inline google-cloud-compute/src/main/java/com/google/cloud/compute/testing/ManagedResourceHelper.java]: done

File: google-cloud-compute/src/main/java/com/google/cloud/compute/testing/ManagedResourceHelper.java (hunk 6 above)
Comment (neozwu) [inline google-cloud-compute/src/main/java/com/google/cloud/compute/testing/ManagedResourceHelper.java]: now use overloading methods for all supported type, so we know whether it's supported at compile time.

File: google-cloud-compute/src/test/java/com/google/cloud/compute/it/ITComputeTest.java [hunk 11]
```diff
@@ -1155,9 +1193,13 @@ public void testListDisksAndSnapshots() throws InterruptedException, TimeoutExce
     }
     assertEquals(2, count);
     compute.deleteDisk(firstDiskId);
+    managedResourceCleaner.remove(firstDiskId);
     compute.deleteDisk(secondDiskId);
+    managedResourceCleaner.remove(secondDiskId);
     compute.deleteSnapshot(firstSnapshotId);
+    managedResourceCleaner.remove(firstSnapshotId);
     compute.deleteSnapshot(secondSnapshotId);
+    managedResourceCleaner.remove(secondSnapshotId);
```
Comment (garrettjonesgoogle) [inline google-cloud-compute/src/test/java/com/google/cloud/compute/it/ITComputeTest.java]: For tests where there is no check after deletion, you could just remove the deletes altogether, since the resource cleaner will take care of it in `tearDown`.

**Comment by neozwu** [comment]
@shinfan @garrettjonesgoogle PTAL

File: google-cloud-compute/src/main/java/com/google/cloud/compute/testing/ManagedResourceHelper.java (hunk 1 above)
Comment (garrettjonesgoogle) [inline google-cloud-compute/src/main/java/com/google/cloud/compute/testing/ManagedResourceHelper.java]: I agree about just focusing on Compute to start with.

File: google-cloud-compute/src/main/java/com/google/cloud/compute/testing/ManagedResourceCleaner.java (hunk 10 above)
Comment (neozwu) [inline google-cloud-compute/src/main/java/com/google/cloud/compute/testing/ManagedResourceCleaner.java]: done

File: google-cloud-compute/src/main/java/com/google/cloud/compute/testing/ManagedResourceCleaner.java (hunk 9 above)
Comment (neozwu) [inline google-cloud-compute/src/main/java/com/google/cloud/compute/testing/ManagedResourceCleaner.java]: done, although still have type parameterization.

File: google-cloud-compute/src/test/java/com/google/cloud/compute/it/ITComputeTest.java (hunk 11 above)
Comment (neozwu) [inline google-cloud-compute/src/test/java/com/google/cloud/compute/it/ITComputeTest.java]: done

File: google-cloud-compute/src/main/java/com/google/cloud/compute/testing/ResourceCleaner.java [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+    this.compute = compute;
+    this.resources = new HashMap<>();
+  }
+
+  public static ResourceCleaner create(Compute compute) {
+    return new ResourceCleaner(compute);
+  }
+
+  private class ManagedResource<T> {
+    final T res;
```
Comment (garrettjonesgoogle) [inline google-cloud-compute/src/main/java/com/google/cloud/compute/testing/ResourceCleaner.java]: rename: `resourceId` (this variable holds an id, not a full resource object)

File: google-cloud-compute/src/main/java/com/google/cloud/compute/testing/ResourceCleaner.java [hunk 13]
```diff
@@ -0,0 +1,179 @@
...
+    Function<T, Void> func;
```
Comment (garrettjonesgoogle) [inline google-cloud-compute/src/main/java/com/google/cloud/compute/testing/ResourceCleaner.java]: rename: `deleteFunc` (`func` is much too general)

File: google-cloud-compute/src/main/java/com/google/cloud/compute/testing/ResourceCleaner.java [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+    }
+
+    public void delete() {
+      func.apply(res);
+    }
+  }
+
+  public void cleanUp() {
+    for (Map.Entry<Object, ManagedResource<?>> r : resources.entrySet()) {
+      System.out.println("Clean up compute engine resource: " + r.getKey().toString());
```
Comment (garrettjonesgoogle) [inline google-cloud-compute/src/main/java/com/google/cloud/compute/testing/ResourceCleaner.java]: I don't think we need to commit this println.

File: google-cloud-compute/src/main/java/com/google/cloud/compute/testing/ResourceCleaner.java [hunk 15]
```diff
@@ -0,0 +1,179 @@
...
+      r.getValue().delete();
+    }
+    resources.clear();
+  }
+
+  public ResourceCleaner add(AddressId addressId) {
```
Comment (garrettjonesgoogle) [inline google-cloud-compute/src/main/java/com/google/cloud/compute/testing/ResourceCleaner.java]: I think we should remove the return value from the `add` methods since it's not used (and I don't see a use for it).

File: google-cloud-compute/src/main/java/com/google/cloud/compute/testing/ResourceCleaner.java [hunk 16]
```diff
@@ -0,0 +1,179 @@
...
+              public Void apply(InstanceId instanceId) {
+                compute.deleteInstance(instanceId);
+                return null;
+              }
+            });
+    resources.put(instanceId, r);
+    return this;
+  }
+
+  public <T> ResourceCleaner remove(T resourceId) {
```
Comment (garrettjonesgoogle) [inline google-cloud-compute/src/main/java/com/google/cloud/compute/testing/ResourceCleaner.java]: The type parameterization isn't necessary here - you can just accept `Object`.

File: google-cloud-compute/src/test/java/com/google/cloud/compute/it/ITComputeTest.java [hunk 17]
```diff
@@ -109,6 +111,12 @@
   public static void beforeClass() {
     RemoteComputeHelper computeHelper = RemoteComputeHelper.create();
     compute = computeHelper.getOptions().getService();
+    resourceCleaner = ResourceCleaner.create(compute);
+  }
+
+  @AfterClass
+  public static void teardown() {
```
Comment (garrettjonesgoogle) [inline google-cloud-compute/src/test/java/com/google/cloud/compute/it/ITComputeTest.java]: Make the 'd' here upper case to stop codacy from yelling at you.

File: google-cloud-compute/src/main/java/com/google/cloud/compute/testing/ResourceCleaner.java (hunk 12 above)
Comment (neozwu) [inline google-cloud-compute/src/main/java/com/google/cloud/compute/testing/ResourceCleaner.java]: done

File: google-cloud-compute/src/main/java/com/google/cloud/compute/testing/ResourceCleaner.java (hunk 13 above)
Comment (neozwu) [inline google-cloud-compute/src/main/java/com/google/cloud/compute/testing/ResourceCleaner.java]: done

File: google-cloud-compute/src/main/java/com/google/cloud/compute/testing/ResourceCleaner.java (hunk 14 above)
Comment (neozwu) [inline google-cloud-compute/src/main/java/com/google/cloud/compute/testing/ResourceCleaner.java]: the println is deleted

File: google-cloud-compute/src/main/java/com/google/cloud/compute/testing/ResourceCleaner.java (hunk 15 above)
Comment (neozwu) [inline google-cloud-compute/src/main/java/com/google/cloud/compute/testing/ResourceCleaner.java]: done

File: google-cloud-compute/src/main/java/com/google/cloud/compute/testing/ResourceCleaner.java (hunk 16 above)
Comment (neozwu) [inline google-cloud-compute/src/main/java/com/google/cloud/compute/testing/ResourceCleaner.java]: done

File: google-cloud-compute/src/test/java/com/google/cloud/compute/it/ITComputeTest.java (hunk 17 above)
Comment (neozwu) [inline google-cloud-compute/src/test/java/com/google/cloud/compute/it/ITComputeTest.java]: done

**Comment by garrettjonesgoogle** [comment]
LGTM
